clippy_lints/src/paths_from_format.rs

```diff
@@ -0,0 +1,124 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::macros::{root_macro_call, FormatArgsExpn};
+use clippy_utils::sugg::Sugg;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+use std::fmt::Write as _;
+use std::path::Path;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `PathBuf::from(format!(..))` calls.
+    ///
+    /// ### Why is this bad?
+    /// It is not OS-agnostic, and can be harder to read.
+    ///
+    /// ### Known Problems
+    /// `.join()` introduces additional allocations that are not present when `PathBuf::push` is
+    /// used instead.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::path::PathBuf;
+    /// let base_path = "/base";
+    /// PathBuf::from(format!("{}/foo/bar", base_path));
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::path::Path;
+    /// let base_path = "/base";
+    /// Path::new(&base_path).join("foo").join("bar");
+    /// ```
+    #[clippy::version = "1.62.0"]
+    pub PATHS_FROM_FORMAT,
+    pedantic,
+    "builds a `PathBuf` from a format macro"
+}
+
+declare_lint_pass!(PathsFromFormat => [PATHS_FROM_FORMAT]);
+
+impl<'tcx> LateLintPass<'tcx> for PathsFromFormat {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if_chain! {
+            if let ExprKind::Call(_, args) = expr.kind;
+            if let ty = cx.typeck_results().expr_ty(expr);
+            if is_type_diagnostic_item(cx, ty, sym::PathBuf);
+            if !args.is_empty();
+            if let Some(macro_call) = root_macro_call(args[0].span);
+            if cx.tcx.item_name(macro_call.def_id) == sym::format;
+            if let Some(format_args) = FormatArgsExpn::find_nested(cx, &args[0], macro_call.expn);
```

Comment:
I know there has been quite some work, on format macro diagnostics, since this PR. Hopefully, a rebase is all we need :)